src/Database/LSMTree/Internal/Run/Reader.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE BangPatterns    #-}
+{-# LANGUAGE LambdaCase      #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | A run that is being read incrementally.
+--
+module Database.LSMTree.Internal.Run.Reader (
+    Reader
+  , new
+  , next
+  , Result (..)
+  , BlobRef (..)
+  , close
+  ) where
+
+import           Control.Exception (handleJust)
+import           Control.Monad (guard)
+import           Control.Monad.Primitive (RealWorld)
+import           Data.IORef
+import           Data.Maybe (fromMaybe)
+import           Data.Primitive.ByteArray (newPinnedByteArray,
+                     unsafeFreezeByteArray)
+import           Data.Primitive.PrimVar
+import           Data.Word (Word16, Word32)
+import           Database.LSMTree.Internal.BitMath (roundUpToPageSize)
+import           Database.LSMTree.Internal.BlobRef (BlobSpan (..))
+import           Database.LSMTree.Internal.Entry
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.RawOverflowPage (RawOverflowPage,
+                     rawBytesToOverflowPages)
+import           Database.LSMTree.Internal.RawPage
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Run.FsPaths
+import           Database.LSMTree.Internal.Serialise
+import qualified System.FS.API as FS
+import           System.FS.API (HasBufFS, HasFS)
+
+-- | Allows reading the k\/ops of a run incrementally, using its own read-only
+-- file handle and in-memory cache of the current disk page.
+--
+-- Creating a 'Reader' does not increase the run's reference count, so make sure
+-- the run remains open while using the 'Reader'.
```

Comment:
IIRC should increase the reference count, or we won't know when it's safe to clean up a run

Reply:
(discussed separately)